libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp

```diff
@@ -155,56 +114,119 @@ void SortitionParamsManager::pbftBlockPushed(const SyncBlock& block, DbStorage::
   if (config_.changing_interval == 0) {
     return;
   }
-
-  uint16_t dag_efficiency = calculateDagEfficiency(block);
-  dag_efficiencies_.push_back(dag_efficiency);
-  const auto& period = block.pbft_blk->getPeriod();
-  db_->savePbftBlockDagEfficiency(period, dag_efficiency, batch);
-  LOG(log_dg_) << period << " pbftBlockPushed, efficiency: " << dag_efficiency / 100. << "%";
-
-  if (non_empty_pbft_chain_size % config_.changing_interval == 0) {
-    const auto params_change = calculateChange(non_empty_pbft_chain_size);
-    if (params_change) {
-      db_->saveSortitionParamsChange(period, *params_change, batch);
-      params_changes_.push_back(*params_change);
+  if (ignored_efficiency_counter_ >= config_.computation_interval - config_.changing_interval) {
+    uint16_t dag_efficiency = calculateDagEfficiency(block);
+    dag_efficiencies_.push_back(dag_efficiency);
+    const auto& period = block.pbft_blk->getPeriod();
+    LOG(log_dg_) << period << " pbftBlockPushed, efficiency: " << dag_efficiency / 100. << "%";
+
+    if (non_empty_pbft_chain_size % config_.changing_interval == 0) {
+      const auto params_change = calculateChange(period);
+      params_changes_.push_back(params_change);
+      db_->saveSortitionParamsChange(period, params_change, batch);
+      cleanup();
     }
+  } else {
+    ignored_efficiency_counter_++;
+  }
+}
 
-    cleanup(period);
+int32_t efficiencyToChange(uint16_t efficiency, uint16_t goal_efficiency) {
+  uint16_t deviation = std::abs(efficiency - goal_efficiency) * 100 / goal_efficiency;
+  // If goal is 50 % return 1% change for 40%-60%, 2% change for 30%-70% and 5% change for over that
+  if (deviation < 20) {
+    return UINT16_MAX / 100;
   }
+  if (deviation < 40) {
+    return UINT16_MAX / 50;
+  }
+  return UINT16_MAX / 20;
 }
 
-int32_t SortitionParamsManager::getChange(uint64_t period, uint16_t efficiency) const {
-  const uint16_t per_percent = averageCorrectionPerPercent();
-  int32_t correction = -(config_.targetEfficiency() - efficiency);
-  if (std::abs(correction) > config_.max_interval_correction) {
-    correction = (correction < 0 ? -1 : 1) * config_.max_interval_correction;
+int32_t SortitionParamsManager::getNewUpperRange(uint16_t efficiency) const {
+  if (efficiency >= config_.dag_efficiency_targets.first && efficiency <= config_.dag_efficiency_targets.second) {
+    return params_changes_[params_changes_.size() - 1].vrf_params.threshold_upper;
```

Comment:
params_changes_ can it be empty?

Reply:
Same here, this is never called with empty params_changes_. I added an assert.